content/blog/2020-05-dplyr-1-0-0-last-minute-additions/index.Rmarkdown

```diff
@@ -0,0 +1,147 @@
+-   [Notes for package developers](https://www.tidyverse.org/blog/2020/04/dplyr-1-0-0-package-dev/)
+
+Today I wanted to talk about two cool new features that we've added since I started blogging about dplyr 1.0.0: `summarise()` now gives you greater control over how the results and grouped, and we've added a new set of functions for modifying rows.
+
+### Getting the dev version
+
+If you'd like to try out anything you see in this blog post, you can install the development version of dplyr with:
+
+```{r, eval = FALSE}
+devtools::install_github("tidyverse/dplyr")
+```
+
+Note that the development version won't become 1.0.0 until it's released, but at this point, it's very similar to what we'll be submitting to CRAN on May 15.
+
+```{r setup}
+library(dplyr, warn.conflicts = FALSE)
+```
+
+## `summarise()` and grouping
+
+There\'s a common confusion about the result of `summarise()`. How do you think the result of the following code will be grouped?
+
+```{r, message = FALSE}
+homeworld_species <- starwars %>% 
+  group_by(homeworld, species) %>% 
+  summarise(n = n())
+```
+
+You might be surprised to learn that it's grouped by `homeworld`:
+
+```{r}
+head(homeworld_species, 3)
+```
+
+That's because `summarise()` always peels off the last group, based on the logic that this group now occupies a single row so there's no point grouping by it. This behaviour made perfect sense to me at the time I implemented it, but it's been a long standing source of confusion amongst dplyr users (and it doesn't make sense if your summary [returns multiple rows](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-summarise/)).
+
+Unfortunately, it would be very difficult to change this default now because a lot of code probably relies on it. Instead, we're doing the next best thing: exposing the default behaviour more explicit and making it easier to change. In dplyr 1.0.0, the code above will display a message telling you how the result has been grouped:
+
+```{r}
+homeworld_species <- starwars %>% 
+  group_by(homeworld, species) %>% 
+  summarise(n = n())
+```
+
+The text hints at how to take control of grouping and eliminate the message: a new `.groups` argument allows you to control the grouping of the result. It currently has four possible values:
+
+-   `.groups = "drop_last"` drops the last grouping level (i.e. the default behaviour sans message).
+-   `.groups = "drop"` drops all grouping levels and returns a tibble.
+-   `.groups = "keep"` preserves the grouping of the input.
+-   `.groups = "rowwise"` turns each row into its own group.
+
+If you find the default message annoying you can suppress by setting a global option:
+
+```{r, eval = FALSE}
+options(dplyr.summarise.inform = FALSE)
+```
+
+`.groups` is very new, so we've marked it as experimental, meaning that it may change in the future. Please let us know what you think of it so that we have enough data to make a decision about its future.
```

Comment:
```suggestion
`.groups` is very new, so we've marked it as experimental, meaning that it may change in the future. Please let us know what you think of it to help us make a decision about its future.
```

Reply:
My underlying principle for the above change is that you never want to imply that feedback has more weight than it does. I think "…so that we have enough data…" suggests we don't make decisions until there is a critical mass of feedback.